[AIEX] Extend Staged 2D/3D regalloc to avoid spills

andcarminati opened the pull request:
This work is intended to avoid 2D/3D (when possible) register spills.

The idea and rationale behind this work is in a previous Draft PR: https://github.com/Xilinx/llvm-aie/pull/442.

To review, I recommend to follow this PR commit by commit.

Credits also for the co-author @krishnamtibrewala.

andcarminati committed:
[AIEX][NFC] Refator some AIESuperReg utilities

Co-Authored-By: Krishnam Tibrewala <krishnam.tibrewala@amd.com>

andcarminati committed:
[AIE2P] Extend allocation filter to include only Target MIs

Now we filter by register class and usage. Basically, we exclude here
instructions like copies and non-2D/3D ones.

Co-Authored-By: Krishnam Tibrewala <krishnam.tibrewala@amd.com>

andcarminati committed:
[AIEX] Fix LIs in SuperRegRewriter that are related to unallocated registers

Co-Authored-By: Krishnam Tibrewala <krishnam.tibrewala@amd.com>

andcarminati committed:
[AIEX] Add a Pass to expand unallocated 2D/3D into individual ones

If we don't need a full register, we can expand to individual lanes.

Co-Authored-By: Krishnam Tibrewala <krishnam.tibrewala@amd.com>

andcarminati committed:
[AIEX] Expand copy bundles for unallocated dest VRegs

This avoids cycles in bundles that appear in VirtRegRewriter.
We also update LIs related to src and dst operands of those
expanded copies.

Co-Authored-By: Krishnam Tibrewala <krishnam.tibrewala@amd.com>

andcarminati committed:
[AIE2P] Add base tests exposing cycles in copy bundles

Co-Authored-By: Krishnam Tibrewala <krishnam.tibrewala@amd.com>

andcarminati committed:
[AIE2P] Add a base lit test with unallocated 2D/3D regs before main Greedy run

Co-Authored-By: Krishnam Tibrewala <krishnam.tibrewala@amd.com>

andcarminati commented:
QoR results:

<img width="4400" height="1800" alt="Core_Insn_Count" src="https://github.com/user-attachments/assets/fd8a85b8-48d5-439e-a985-6975060ad03e" />
<img width="4400" height="1800" alt="Core_StackSize_absolute" src="https://github.com/user-attachments/assets/8a31c55c-94ec-41a1-9d2d-862603782f72" />
<img width="4400" height="1800" alt="Core_PMSize_absolute" src="https://github.com/user-attachments/assets/624fe2d6-f73e-4f59-ad94-6e149b2c48e0" />

krishnamtibrewala commented:
Thanks you @andcarminati, very much !!

krishnamtibrewala commented on llvm/lib/Target/AIE/aie2p/AIE2PTargetMachine.cpp:
```diff
@@ -124,6 +124,8 @@ bool AIE2PPassConfig::addRegAssignAndRewriteOptimized() {
     addPass(createGreedyRegisterAllocator(onlyAllocate3D2DRegisters));
     addPass(createAIESuperRegRewriter());
+    if (EnableFineGrainedStagedRA)
+      addPass(createAIEUnallocatedSuperRegRewriter());
```
Check : I think we should try the following, it should be helpful for `onlyAllocate3D2DRegisters` pass

```
    addPass(createGreedyRegisterAllocator(onlyAllocate3DRegisters));
    addPass(createAIESuperRegRewriter());
    if (EnableFineGrainedStagedRA)
      addPass(createAIEUnallocatedSuperRegRewriter());
    addPass(createGreedyRegisterAllocator(onlyAllocate3D2DRegisters));
    addPass(createAIESuperRegRewriter());
    if (EnableFineGrainedStagedRA)
      addPass(createAIEUnallocatedSuperRegRewriter());
```

andcarminati replied:
I would like to keeps as simple as possible now to setup a common ground first, and then we can test small changes to see the impact. I think it will better to decide.

krishnamtibrewala replied:
Sounds good!

krishnamtibrewala commented:
Also do you think following commit will help ? 
https://github.com/Xilinx/llvm-aie/pull/442/commits/a681b6ec5cfef563739a68382eb03970d452d487

andcarminati commented:
> Also do you think following commit will help ? [a681b6e](https://github.com/Xilinx/llvm-aie/commit/a681b6ec5cfef563739a68382eb03970d452d487)

Maybe yes! As mentioned before, I prefer to keep just the minimal necessary changes. We can test after, on top of this PR.

F-Stuckmann commented on llvm/lib/Target/AIE/aie2p/AIE2PTargetMachine.cpp:
```diff
@@ -70,17 +72,44 @@ void AIE2PPassConfig::addPreRegBankSelect() {
+  const TargetRegisterClass *RegClass = MRI.getRegClass(R);
+  if (!AIE2P::eDSRegClass.hasSubClassEq(RegClass))
+    return false;
+  return EnableFineGrainedStagedRA ? isRegUsedBy2DOr3DInstruction(MRI, R)
```
CHECK: is the function name still correct, since we also querry 2D registers?

F-Stuckmann replied:
Why do we need need this extra check?

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegRewriter.cpp:
```diff
@@ -232,10 +236,32 @@ static void rewriteFullCopy(MachineInstr &MI, const std::set<int> &CopySubRegs,
+      const LiveInterval &OldLI = LIS.getInterval(R);
+      LRM.unassign(OldLI);
+      LIS.removeInterval(R);
+      const LiveInterval &LI = LIS.getInterval(R);
```
nit: clarify that you create a new Interval here by using: `createAndComputeVirtRegInterval`

andcarminati replied:
I replaced by `createAndComputeVirtRegInterval` to be clear.

andcarminati committed:
[AIE2P] Add a base test to detect missing undef flags in copy expansion.

The goal of this test is to check if we properly insert undef flag on the def side
of a expanded full copy.  On a sub-register def operand, it refers to the part of the
register that isn't written. A sub-register def implicitly reads the other parts of the
register being redefined unless the <undef> flag is set, and a missing flag can
force the related register to be inserted in liveout set of the predecessors block,
causing dominance problems.

Co-Authored-By: Krishnam Tibrewala <krishnam.tibrewala@amd.com>

andcarminati committed:
[AIEX] Add missing undef flags in copy expansion

This will handle properly use of non-dominating definitions. We also
change the handling of the destination registers in two parts:

*Copy expansion: we replace the ogininal index by the index of the first
lane copy to avoid the creation LRs with just one instruction, in this
way we keep que LI correct.

*Rewrite: reset dead flags if necessary.

Co-Authored-By: Krishnam Tibrewala <krishnam.tibrewala@amd.com>

F-Stuckmann commented on llvm/lib/Target/AIE/aie2p/AIE2PTargetMachine.cpp:
```diff
@@ -70,17 +72,44 @@ void AIE2PPassConfig::addPreRegBankSelect() {
+            MI.getMF()->getSubtarget().getInstrInfo());
+
+        // We should recognize both cases, with and without splitting. A 2D/3D
+        // instruction will always be split os splittable.
```
nit: as

andcarminati replied:
Repaired as a fixup in the refactor commit.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegRewriter.cpp:
```diff
@@ -232,10 +236,32 @@ static void rewriteFullCopy(MachineInstr &MI, const std::set<int> &CopySubRegs,
   MI.eraseFromParent();
+  // As we don't handle all registers now (selective LI filter),
+  // We should make sure that all LiveIntervals are correct.
+  // If we dont't repair, MI will compose the LIs of some registers,
```
nit: don't

F-Stuckmann replied:
What do you mean with compose? the dead MI will block some LI ?

andcarminati replied:
A dead MI will have an invalid SlotIndex that will appear in some instruction`s LI, what is wrong. The original copy should not appear in any LI.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.h:
```diff
@@ -0,0 +1,72 @@
+                                      const AIEBaseRegisterInfo &TRI,
+                                      std::set<Register> &VisitedVRegs);
+
+SmallSet<int, 8> getRewritableSubRegs(Register Reg,
```
nit: can we have a comment what this function does?

andcarminati replied:
Yes, it is a refactor but it is always good to document.

F-Stuckmann replied:
Could we swap the documentation here? 
I think the high level function should get the documentation, the actual implementation function (the one above this) should get the slimmed documentation.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,265 @@
+      }
+      Register SrcReg = RegOp.getParent()->getOperand(1).getReg();
+      if (!VisitedVRegs.count(SrcReg) &&
+          getRewritableSubRegs(SrcReg, MRI, TRI, VisitedVRegs).empty()) {
```
Does it ever happen, that SrcReg has no SubRegs, but DstReg has them or vis versa? 
Can they also have different SubRegs?

andcarminati replied:
This part was a refactor, but as we are handling a full copy here, we can expect subregs on both sides.

andcarminati replied:
As a full copy, I mean a 2d to 2d copy or a 3d to 3d copy.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,265 @@
+  MI.eraseFromParent();
+  // As we don't handle all registers now (selective LI filter),
+  // We should make sure that all LiveIntervals are correct.
+  // If we dont't repair, MI will compose the LIs of some registers,
```
nit: don't

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,265 @@
+  // We should make sure that all LiveIntervals are correct.
+  // If we dont't repair, MI will compose the LIs of some registers,
+  // what is not correct because MI was deleted.
+  updateLRMAndLIS(RegistersToRepair, VRM, LRM, LIS);
```
nit: could we get a better name that does not have the abbreviations in it? 
`updateLiveIntervals`. Also since LIS is the most relevant Datastructure, have it first and then followed by helper structures `LRM` and `VRM`.

andcarminati commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -103,9 +103,9 @@ void rewriteFullCopy(MachineInstr &MI, const std::set<int> &CopySubRegs,
-
+  MachineInstr *FirstMI = nullptr;
   SmallSet<Register, 8> RegistersToRepair;
+  bool IsFirstCopy = true;
```
nit: We can use FirstMI as a replacement for IsFirstCopy.

andcarminati commented on llvm/lib/Target/AIE/AIEUnallocatedSuperRegRewriter.cpp:
```diff
@@ -0,0 +1,172 @@
+//===----- AIEUnallocatedSuperRegRewriter.cpp - Constrain tied sub-registers
```
Fix this.

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/ra/staged-ra-check-undef.mir:
```diff
@@ -0,0 +1,164 @@
+# RUN: llc -O2 -mtriple=aie2p -start-before=greedy \
+# RUN:     -stop-before=aie-unallocated-superreg-rewrite %s -o - | FileCheck %s
+
+# The goal of this test is to check if we properly insert undef flag on the def side
```
I could extend the verifier to cover partial definitions. It is far away from be trivial, but it could be done.

martien-de-jong commented on llvm/lib/Target/AIE/aie2p/AIE2PTargetMachine.cpp:
```diff
@@ -60,17 +62,44 @@ void AIE2PPassConfig::addPreRegBankSelect() {
+static bool isRegUsedBy2DOr3DInstruction(const MachineRegisterInfo &MRI,
+                                         const Register &R) {
+
+  return std::any_of(
```
llvm::any_of on the range?

martien-de-jong commented on llvm/lib/Target/AIE/AIEUnallocatedSuperRegRewriter.cpp:
```diff
@@ -125,6 +126,62 @@ void rewriteCandidates(RegRewriteInfo &RegistersToRewrite,
+
+    for (MachineInstr &MI : MRI.def_instructions(VReg)) {
+
+      if (!MI.isBundledWithPred() || MI.isBundledWithSucc())
```
Can I have a comment explaining this directional asymmetry?

andcarminati commented:
There is no failures across all benchmarks.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+
+    MachineInstr *PartCopy = BuildMI(*MI.getParent(), MI, MI.getDebugLoc(),
+                                     TII.get(TargetOpcode::COPY))
+                                 .addReg(DstReg, RegState::Define, SubRegIdx)
```
nit: we do not need undefs here because we fully define the subregisters that we copy to?

andcarminati replied:
Yes, we start the definition on the first copy, and we consider that next copies will read the previous one. In practice, inserting undefs in all copies will not hurt in general, but as is, we have a more accurate model for incremental definitions.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+
+    // Only set undefined on the first partial copy. The first copy doesn't read
+    // other lanes, but subsequent copies do read the previously written lanes.
+    // Setting undefined on all copies breaks live interval tracking and causes
```
oh yeah i also found that out with subreg spilling XD

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+
+  // Replace the original copy by the first one, so we automatically repair
+  // DstReg's LI.
+  LIS.ReplaceMachineInstrInMaps(MI, *FirstMI);
```
why aren't we inserting all the newly created machineinstructions into LIS?

andcarminati replied:
Because, later on, each new lane copy will make dead the previous one if we try to fix the LI. In this case, we say that the LR will start on the first copy and then it will simply continue though the next ones, until the next last use. I this way we prevent false dead flags everywhere.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+}
+
+/// Rewrite a full copy into multiple copies using the subregs in \p CopySubRegs
+void rewriteFullCopy(MachineInstr &MI, const std::set<int> &CopySubRegs,
```
MI is a copy, right? Could you rename MI to CopyMI, so it becomes clearer?

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+  for (MachineInstr &MI : make_early_inc_range(MRI.reg_instructions(Reg))) {
+    if (MI.isFullCopy())
+      AIESuperRegUtils::rewriteFullCopy(
+          MI, TRI.getSubRegSplit(MRI.getRegClass(Reg)->getID()), LIS, *TII, TRI,
```
you are also passing TRI, I think you don't need to pass CopySubRegs here

You can get MRI like this: 
MachineRegisterInfo &MRI = MI.getMF()->getRegInfo();

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+  // We should make sure that all LiveIntervals are correct.
+  // If we don't repair, MI will compose the LIs of some registers,
+  // what is not correct because MI was deleted.
+  repairLiveIntervals(RegistersToRepair, VRM, LRM, LIS);
```
Can't you use
LIS.InsertMachineInstrRangeInMaps(AllSubRegCopies)?

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+}
+
+/// Return a mask of all the lanes that are live at \p Index
+LaneBitmask getLiveLanesAt(SlotIndex Index, Register Reg,
```
I love this function! Thank you very much

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+                     SlotIndexes &Indexes, LiveDebugVariables &DebugVars) {
+  LLVM_DEBUG(dbgs() << "Rewriting " << printReg(Reg, &TRI, 0, &MRI) << '\n');
+  MachineFunction &MF = VRM.getMachineFunction();
+  auto *TII =
```
nit: const

F-Stuckmann commented on llvm/lib/Target/AIE/AIEUnallocatedSuperRegRewriter.cpp:
```diff
@@ -0,0 +1,235 @@
+/// Split candidate registers into independent virtual registers for each
+/// subregister. Each composite register is rewritten using its subregister
+/// indices, with live intervals and debug information updated accordingly.
+void rewriteCandidates(RegRewriteInfo &RegistersToRewrite,
```
Are these the candidates to rewrite? 
nit: rename for clarity

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+  assert(!SubRegs.empty());
+  for (int SubReg : SubRegs) {
+    const TargetRegisterClass *SubRC =
+        AssignedPhysReg.has_value()
```
why do we care if the AssignedPhysReg is passed? 
If we go for largestLegalSuperclass, we aren't really going for subreg rewriting, but for a regular copy.

F-Stuckmann replied:
Shouldn't the check be if a subregisterClass exists and if we fail to retrieve it we should go for largestlegalSuperclass? 
In the failure to get a subregClass should we still iterate through all the subregs? 
Don't group multiple together if we use a superclass?

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+
+  // Rewrite full copies into multiple copies using subregs
+  for (MachineInstr &MI : make_early_inc_range(MRI.reg_instructions(Reg))) {
+    if (MI.isFullCopy())
```
Does `reg_instructions(Reg)` also return subreg uses? 
If not at this point we can assume that Reg is a superregister and exit early if we aren't a superregister.

F-Stuckmann replied:
yeah it does.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+  LLVM_DEBUG(dbgs() << "  Splitting range " << LIS.getInterval(Reg) << "\n");
+  for (MachineOperand &RegOp : make_early_inc_range(MRI.reg_operands(Reg))) {
+    LLVM_DEBUG(dbgs() << "  Changing " << *RegOp.getParent());
+    int SubReg = RegOp.getSubReg();
```
How do we know, we don't have further subregs below us?

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+    // TODO: One could use collectSubRegs() in AIEBaseInstrInfo.cpp
+    // But given that MOD registers are not part of the ABI, they should
+    // not appear as physical registers before RA.
+    LLVM_DEBUG(dbgs() << "  Cannot rewrite physreg " << printReg(Reg, &TRI)
```
nit message should make it clear, that we cannot collect subregs to rewrite, since it already physical

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+  VisitedVRegs.insert(Reg);
+  SmallSet<int, 8> UsedSubRegs;
+  for (MachineOperand &RegOp : MRI.reg_operands(Reg)) {
+    int SubReg = RegOp.getSubReg();
```
nit: const

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+  SmallSet<int, 8> UsedSubRegs;
+  for (MachineOperand &RegOp : MRI.reg_operands(Reg)) {
+    int SubReg = RegOp.getSubReg();
+    if (SubReg && SubRegSplit.count(SubReg)) {
```
if we are a subreg, shouldn't we automatically be in the SubRegSplit? 
I think we can assert it instead of checking it in the condition.

andcarminati replied:
I will keep as is, this code was just refactored.

F-Stuckmann replied:
Adding an assert crashes here because we cannot properly handle low and high parts of the 3d Registers, so a whole d0 or d4 register in the case of 3d_0.
They do not get returned from the `TRI.getSubRegSplit()`
Could you add a comment ? 
Also do we know, why we did not add it to `TRI.getSubRegSplit()`?

F-Stuckmann replied:
These are not full copies but partial copies, that we currently don't handle

F-Stuckmann replied:
I think we are leaving something on the table here. not sure if it is needed for this PR though

andcarminati replied:
> Also do we know, why we did not add it to TRI.getSubRegSplit()?
We always rewrite into the smallest lanes as possible.

Here we are concerned with the things that we can rewrite after. Any update here should be aligned with the rewrite after, and it is not clear the benefits afterwards. I really don't want to complicate this PR even more. Any improvement could be done with an established baseline as follow up.

andcarminati replied:
I included a TODO comment related to this investigation.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+
+  VisitedVRegs.insert(Reg);
+  SmallSet<int, 8> UsedSubRegs;
+  for (MachineOperand &RegOp : MRI.reg_operands(Reg)) {
```
Maybe add a comment that we are walking all defs and uses to find subreg uses and full copies.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+    int SubReg = RegOp.getSubReg();
+    if (SubReg && SubRegSplit.count(SubReg)) {
+      UsedSubRegs.insert(SubReg);
+    } else if (RegOp.getParent()->isFullCopy()) {
```
Why are we performing this check on uses? 
Isn't this a check we can perform once on the DefI?

andcarminati replied:
This was just a refactor.

F-Stuckmann replied:
Is there a reason why we don't handle copies uniform in this method? 
we also encounter copies in this case here `if (SubReg && SubRegSplit.count(SubReg))`

andcarminati replied:
Because we should be able to full expand the copy after. At this moment we don't handle cases like `%7.sub_hi_dim:eds = COPY %6:ed`.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,321 @@
+      }
+      UsedSubRegs.insert(SubRegSplit.begin(), SubRegSplit.end());
+    } else {
+      LLVM_DEBUG(dbgs() << "  Cannot rewrite " << RegOp << " in "
```
Is this a regular use/Def of the full register?

andcarminati replied:
This was just a refactor.

andcarminati replied:
This will catch things like `%7.sub_hi_dim:eds = COPY %6:ed`, not handled after.

andcarminati commented on llvm/lib/Target/AIE/AIESuperRegRewriter.cpp:
```diff
@@ -232,10 +236,32 @@ static void rewriteFullCopy(MachineInstr &MI, const std::set<int> &CopySubRegs,
     LLVM_DEBUG(dbgs() << "        to " << *PartCopy);
     LIS.InsertMachineInstrInMaps(*PartCopy);
     LIS.getInterval(PartCopy->getOperand(0).getReg());
+    RegistersToRepair.insert(PartCopy->getOperand(1).getReg());
```
Comment about repair (only r)

andcarminati replied:
The message is in a propriated commit.

andcarminati commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -118,13 +136,31 @@ void rewriteFullCopy(MachineInstr &MI, const std::set<int> &CopySubRegs,
 
-  LIS.RemoveMachineInstrFromMaps(MI);
+  // Replace the original copy by the first one, so we automatically repair
+  // DstReg's LI.
```
Why?

andcarminati commented on llvm/lib/Target/AIE/AIEUnallocatedSuperRegRewriter.cpp:
```diff
@@ -0,0 +1,171 @@
+       ++VRegIdx) {
+    const Register Reg = Register::index2VirtReg(VRegIdx);
+
+    // Ignore un-used registers
```
Complete the comment.

andcarminati commented on llvm/lib/Target/AIE/AIEUnallocatedSuperRegRewriter.cpp:
```diff
@@ -0,0 +1,171 @@
+    if (MRI.reg_nodbg_empty(Reg) || VRM.hasPhys(Reg))
+      continue;
+
+    SmallSet<int, 8> RewritableSubRegs =
```
const

andcarminati commented on llvm/lib/Target/AIE/AIEUnallocatedSuperRegRewriter.cpp:
```diff
@@ -0,0 +1,171 @@
+    LLVM_DEBUG(dbgs() << "  Rewriting " << printReg(VReg, &TRI, 0, &MRI)
+                      << " into " << SubRegs.size() << " subregister(s)\n");
+    AIESuperRegUtils::rewriteSuperReg(
+        VReg, /*std::optional<Register> AssignedPhysReg = */ {}, SubRegs, MRI,
```
Local var for Assign...

andcarminati commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -184,21 +184,26 @@ LaneBitmask getLiveLanesAt(SlotIndex Index, Register Reg,
+    const TargetRegisterClass *SubRC =
+        AssignedPhysReg.has_value()
+            ? TRI.getSubRegisterClass(SuperRC, SubReg)
+            : TRI.getLargestLegalSuperClass(
```
Why?

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/ra/staged-ra-spill.mir:
```diff
@@ -6,42 +6,68 @@
+    ; RA-STAGEG-FG-LABEL: name: test_spill_2d_last_stage
+    ; RA-STAGEG-FG: liveins: $d1, $d2, $d3, $d4, $d5, $d6, $d7, $p0, $p1
+    ; RA-STAGEG-FG-NEXT: {{  $}}
+    ; RA-STAGEG-FG-NEXT: renamable $dn0 = LDA_dms_lda_idx_imm renamable $p1, 0
```
STAGED

andcarminati commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -257,10 +261,13 @@ void rewriteSuperReg(Register Reg, Register AssignedPhysReg,
-      LRM.assign(*LI, SubPhysReg);
-      VRM.setRequiredPhys(LI->reg(), SubPhysReg);
-      LLVM_DEBUG(dbgs() << "  Assigned " << printReg(LI->reg()) << "\n");
+    if (AssignedPhysReg.has_value()) {
```
early return.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,267 @@
+}
+
+/// Rewrite a full copy into multiple copies using the subregs in \p CopySubRegs
+void rewriteFullCopy(MachineInstr &CopyMI, LiveIntervals &LIS,
```
note: in the current state we will never rewrite subreg copies, e.g. copies of d0 or d4 in the case of a 3d_0 register.
Do you know why they are not relevant for the Superregrewriter?

andcarminati replied:
I guess this will never occur because it will be prevented by. `getRewritableSubRegs`.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  // Rewrite full copies into multiple copies using subregs
+  for (MachineInstr &MI : make_early_inc_range(MRI.reg_instructions(Reg))) {
```
Do we rewrite Full copies again when we AssignedPhysReg is not set?

andcarminati replied:
I see that they may occur in case we have a `VReg0_unallocated = COPY VReg1_unallocated`. Considering Conv2D_bft16, we start with  5 3D virtual registers and end up creating 32 ones, so I this may be possible and I would like to keep this this possibility.

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,325 @@
+      rewriteFullCopy(MI, LIS, *TII, TRI, VRM, LRM);
+  }
+
+  LLVM_DEBUG(dbgs() << "  Splitting range " << LIS.getInterval(Reg) << "\n");
```
Do we split ranges again when we AssignedPhysReg is not set?

andcarminati replied:
Yes, in this part of the function we change the proper registers by others with `SubRegisterClass` (allocated) and `LargestLegalSuperClass` (unallocated). Also, we are not doing this again, each pass covesr one subset: allocated or unallocated.

F-Stuckmann commented on llvm/lib/Target/AIE/AIEUnallocatedSuperRegRewriter.cpp:
```diff
@@ -0,0 +1,171 @@
+       ++VRegIdx) {
+    const Register Reg = Register::index2VirtReg(VRegIdx);
+
+    // Ignore un-used od already allocated registers.
```
nit: or

F-Stuckmann commented on llvm/lib/Target/AIE/AIEUnallocatedSuperRegRewriter.cpp:
```diff
@@ -0,0 +1,171 @@
+//
+// (c) Copyright 2025 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
```
What registers aren't assigned after the previous 2d/3d reg allocs? 
Are we changing here the copies and moves not the padds? 
Could you Comment why this pass is necessary and the previous passes do not pick this up?

andcarminati replied:
I will complement this part.

F-Stuckmann commented:
LGTM @martien-de-jong what do you say

F-Stuckmann commented on llvm/lib/Target/AIE/AIESuperRegUtils.cpp:
```diff
@@ -0,0 +1,269 @@
+    int SubReg = RegOp.getSubReg();
+    assert(SubReg);
+    RegOp.setReg(SubRegToVReg[SubReg]);
+    RegOp.setSubReg(0);
```
Note: 3D and 2D reg alloc will have already allocated the the 3D and 2D users except for the copies.
We spill them in line 178, so that we only work on subregs. 
The users we now encounter are the subreg copies from the `rewriteFullCopy`

andcarminati replied:
Please note that `rewriteFullCopy` is to prevent failures during the rewrite. We can rewrite also operands in load and move (also immediate) instructions. See instructions covered by `ItinRegClassPair`.
